### ISSUE TITLE
 Cannot read properties of undefined (reading 'primary')

### ISSUE THREAD

**Issue by cdwmhcc** [directus/directus#20175]
### Describe the Bug

```
npx directus database migrate:latest
```
```bash
[03:24:36.885] WARN: Couldn't load extensions
[03:24:36.886] WARN: Cannot read properties of undefined (reading 'primary')
    err: {
      "type": "TypeError",
      "message": "Cannot read properties of undefined (reading 'primary')",
      "stack":
          TypeError: Cannot read properties of undefined (reading 'primary')
              at getASTFromQuery (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/utils/get-ast-from-query.js:47:56)
              at ItemsService.readByQuery (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/services/items.js:310:25)
              at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
              at async getExtensionsSettings (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/extensions/lib/get-extensions-settings.js:16:22)
              at async ExtensionManager.load (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/extensions/manager.js:138:39)
              at async ExtensionManager.initialize (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/extensions/manager.js:122:13)
              at async createCli (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/cli/index.js:23:9)
    }
[03:24:36.942] INFO: Running migrations...
[03:24:36.949] INFO: Applying Add Content Versioning...
[03:24:36.954] ERROR: create table `directus_versions` (`id` char(36) not null, `key` varchar(64) not null, `name` varchar(255), `collection` varchar(64) not null, `item` varchar(255) not null, `hash` varchar(255), `date_created` timestamp default CURRENT_TIMESTAMP, `date_updated` timestamp default CURRENT_TIMESTAMP, `user_created` char(36), `user_updated` char(36), primary key (`id`)) default character set utf8mb4 - ER_TABLE_EXISTS_ERROR: Table 'directus_versions' already exists
    err: {
      "type": "Error",
      "message": "create table `directus_versions` (`id` char(36) not null, `key` varchar(64) not null, `name` varchar(255), `collection` varchar(64) not null, `item` varchar(255) not null, `hash` varchar(255), `date_created` timestamp default CURRENT_TIMESTAMP, `date_updated` timestamp default CURRENT_TIMESTAMP, `user_created` char(36), `user_updated` char(36), primary key (`id`)) default character set utf8mb4 - ER_TABLE_EXISTS_ERROR: Table 'directus_versions' already exists",
      "stack":
          Error: ER_TABLE_EXISTS_ERROR: Table 'directus_versions' already exists
              at Sequence._packetToError (/apps/web/api/node_modules/.pnpm/mysql@2.18.1/node_modules/mysql/lib/protocol/sequences/Sequence.js:47:14)
              at Query.ErrorPacket (/apps/web/api/node_modules/.pnpm/mysql@2.18.1/node_modules/mysql/lib/protocol/sequences/Query.js:79:18)
              at Protocol._parsePacket (/apps/web/api/node_modules/.pnpm/mysql@2.18.1/node_modules/mysql/lib/protocol/Protocol.js:291:23)
              at Parser._parsePacket (/apps/web/api/node_modules/.pnpm/mysql@2.18.1/node_modules/mysql/lib/protocol/Parser.js:433:10)
              at Parser.write (/apps/web/api/node_modules/.pnpm/mysql@2.18.1/node_modules/mysql/lib/protocol/Parser.js:43:10)
              at Protocol.write (/apps/web/api/node_modules/.pnpm/mysql@2.18.1/node_modules/mysql/lib/protocol/Protocol.js:38:16)
              at Socket.<anonymous> (/apps/web/api/node_modules/.pnpm/mysql@2.18.1/node_modules/mysql/lib/Connection.js:88:28)
              at Socket.<anonymous> (/apps/web/api/node_modules/.pnpm/mysql@2.18.1/node_modules/mysql/lib/Connection.js:526:10)
              at Socket.emit (node:events:517:28)
              at addChunk (node:internal/streams/readable:335:12)
              --------------------
              at Protocol._enqueue (/apps/web/api/node_modules/.pnpm/mysql@2.18.1/node_modules/mysql/lib/protocol/Protocol.js:144:48)
              at Connection.query (/apps/web/api/node_modules/.pnpm/mysql@2.18.1/node_modules/mysql/lib/Connection.js:198:25)
              at /apps/web/api/node_modules/.pnpm/knex@2.4.2_mysql@2.18.1_pg@8.11.0_sqlite3@5.1.6_tedious@16.1.0/node_modules/knex/lib/dialects/mysql/index.js:132:18
              at new Promise (<anonymous>)
              at Client_MySQL._query (/apps/web/api/node_modules/.pnpm/knex@2.4.2_mysql@2.18.1_pg@8.11.0_sqlite3@5.1.6_tedious@16.1.0/node_modules/knex/lib/dialects/mysql/index.js:126:12)
              at executeQuery (/apps/web/api/node_modules/.pnpm/knex@2.4.2_mysql@2.18.1_pg@8.11.0_sqlite3@5.1.6_tedious@16.1.0/node_modules/knex/lib/execution/internal/query-executioner.js:37:17)
              at Client_MySQL.query (/apps/web/api/node_modules/.pnpm/knex@2.4.2_mysql@2.18.1_pg@8.11.0_sqlite3@5.1.6_tedious@16.1.0/node_modules/knex/lib/client.js:146:12)
              at Runner.query (/apps/web/api/node_modules/.pnpm/knex@2.4.2_mysql@2.18.1_pg@8.11.0_sqlite3@5.1.6_tedious@16.1.0/node_modules/knex/lib/execution/runner.js:123:36)
              at Runner.queryArray (/apps/web/api/node_modules/.pnpm/knex@2.4.2_mysql@2.18.1_pg@8.11.0_sqlite3@5.1.6_tedious@16.1.0/node_modules/knex/lib/execution/runner.js:217:21)
              at Runner.queryArray (/apps/web/api/node_modules/.pnpm/knex@2.4.2_mysql@2.18.1_pg@8.11.0_sqlite3@5.1.6_tedious@16.1.0/node_modules/knex/lib/execution/runner.js:269:31)
      "code": "ER_TABLE_EXISTS_ERROR",
      "errno": 1050,
      "sqlMessage": "Table 'directus_versions' already exists",
      "sqlState": "42S01",
      "index": 0,
      "sql": "create table `directus_versions` (`id` char(36) not null, `key` varchar(64) not null, `name` varchar(255), `collection` varchar(64) not null, `item` varchar(255) not null, `hash` varchar(255), `date_created` timestamp default CURRENT_TIMESTAMP, `date_updated` timestamp default CURRENT_TIMESTAMP, `user_created` char(36), `user_updated` char(36), primary key (`id`)) default character set utf8mb4"
    }
```

### To Reproduce

- os: Ubuntu 22.04.3 LTS (GNU/Linux 5.15.0-83-generic x86_64)
- nodejs: 18.18.2
- pnpm: 8.9.2
- directus: 10.7.0

This problem occurred after I upgraded 10.6.3 to 10.7.0. When I run `npx directus start`, it says 'primary' cannot be found. I think I should execute the cli to migrate the database. So I executed `npx directus database migrate:latest` and got the error.

### Directus Version

10.7.0

### Hosting Strategy

Self-Hosted (Custom)

**Comment by rijkvanzanten** [comment]
The `primary` part is actually just a warning, and a duplicate of https://github.com/directus/directus/issues/20131

The error preventing you from running migrate is:

```
[03:24:36.954] ERROR: create table `directus_versions` (`id` char(36) not null, `key` varchar(64) not null, `name` varchar(255), `collection` varchar(64) not null, `item` varchar(255) not null, `hash` varchar(255), `date_created` timestamp default CURRENT_TIMESTAMP, `date_updated` timestamp default CURRENT_TIMESTAMP, `user_created` char(36), `user_updated` char(36), primary key (`id`)) default character set utf8mb4 - ER_TABLE_EXISTS_ERROR: Table 'directus_versions' already exists
```

AKA `directus_versions` already exists.

Did you by chance run the previous `10.7-RC.0` version?

**Comment by cdwmhcc** [comment]
> The `primary` part is actually just a warning, and a duplicate of #20131
> 
> The error preventing you from running migrate is:
> 
> ```
> [03:24:36.954] ERROR: create table `directus_versions` (`id` char(36) not null, `key` varchar(64) not null, `name` varchar(255), `collection` varchar(64) not null, `item` varchar(255) not null, `hash` varchar(255), `date_created` timestamp default CURRENT_TIMESTAMP, `date_updated` timestamp default CURRENT_TIMESTAMP, `user_created` char(36), `user_updated` char(36), primary key (`id`)) default character set utf8mb4 - ER_TABLE_EXISTS_ERROR: Table 'directus_versions' already exists
> ```
> 
> AKA `directus_versions` already exists.
> 
> Did you by chance run the previous `10.7-RC.0` version?

I have never installed 10.7.0 rc.0. I may have executed `npx directus database migrate:latest` this command accidentally. Will it be affected?

**Comment by cdwmhcc** [comment]
I reinstalled 10.6.4 and it worked.I tested `npx directus database migrate:latest` and it seems to have no effect.

``` bash
> npx directus database migrate:latest


   ╭───────────────────────────────────────────────────╮
   │                                                   │
   │                 Update available!                 │
   │                                                   │
   │                  10.6.4 → 10.7.0                  │
   │                 1 version behind                  │
   │                                                   │
   │                 More information:                 │
   │   https://github.com/directus/directus/releases   │
   │                                                   │
   ╰───────────────────────────────────────────────────╯

[04:03:36.790] INFO: Running migrations...
[04:03:36.836] INFO: Applying Update CSV Fields to Text...
[04:03:36.849] INFO: Database up to date
```

``` bash
> npx directus start


   ╭───────────────────────────────────────────────────╮
   │                                                   │
   │                 Update available!                 │
   │                                                   │
   │                  10.6.4 → 10.7.0                  │
   │                 1 version behind                  │
   │                                                   │
   │                 More information:                 │
   │   https://github.com/directus/directus/releases   │
   │                                                   │
   ╰───────────────────────────────────────────────────╯

[04:02:29.810] WARN: PUBLIC_URL should be a full URL
[04:02:29.857] WARN: Database migrations have not all been run
[04:02:29.916] INFO: Watching extensions for changes...
[04:02:29.950] INFO: GraphQL Subscriptions started at ws://0.0.0.0:8055/graphql
[04:02:29.951] INFO: WebSocket Server started at ws://0.0.0.0:8055/websocket
[04:02:29.952] INFO: Server started at http://0.0.0.0:8055
[04:02:43] GET /admin/login?reason=SESSION_EXPIRED 200 5ms
```

**Comment by paescuj** [comment]
`npx directus database migrate:latest` only applies the migrations available in the current version. So in order to see an effect you would have to upgrade to 10.7.0 before.

Not sure how you ended up with the `directus_versions` table existing before 10.7.0 🤔 But to fix it you may want to manually inspect your database, and delete this table if possible.
I recommend doing a backup before that. See also https://docs.directus.io/self-hosted/upgrades-migrations.html.

**Comment by cdwmhcc** [comment]
This table is empty, I can delete it directly, let me try 10.7.0

**Comment by cdwmhcc** [comment]
After deleting the relevant fields and tables, it should be working.

``` bash
> npx directus database migrate:latest

[05:01:36.647] WARN: Couldn't load extensions
[05:01:36.648] WARN: Cannot read properties of undefined (reading 'primary')
    err: {
      "type": "TypeError",
      "message": "Cannot read properties of undefined (reading 'primary')",
      "stack":
          TypeError: Cannot read properties of undefined (reading 'primary')
              at getASTFromQuery (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/utils/get-ast-from-query.js:47:56)
              at ItemsService.readByQuery (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/services/items.js:310:25)
              at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
              at async getExtensionsSettings (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/extensions/lib/get-extensions-settings.js:16:22)
              at async ExtensionManager.load (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/extensions/manager.js:138:39)
              at async ExtensionManager.initialize (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/extensions/manager.js:122:13)
              at async createCli (file:///apps/web/api/node_modules/.pnpm/@directus+api@14.0.0_vue@3.3.4/node_modules/@directus/api/dist/cli/index.js:23:9)
    }
[05:01:36.845] INFO: Loaded extensions: generate-types
[05:01:36.847] INFO: Running migrations...
[05:01:36.856] INFO: Applying Add Content Versioning...
[05:01:37.091] INFO: Applying Themes...
[05:01:37.363] INFO: Applying Update Panel Options...
[05:01:37.369] INFO: Applying Add Extensions...
[05:01:37.384] INFO: Database up to date
```

**Comment by eikaramba** [comment]
i just upgraded from directus 9.24.0 to 10.7.0 and get the same error. however it was just because the directus_migation table was not present at the start of the migration. at the end it was created. not sure if any migrations did fail now. hopefully not

**Comment by Leapfrognz** [comment]
I am also getting this issue. I delete the tables as stated above but then the same thing happens on the next time I migrate. 

`err: {
      "type": "TypeError",
      "message": "Cannot read properties of undefined (reading 'primary')",
      "stack":
          TypeError: Cannot read properties of undefined (reading 'primary')
              at getASTFromQuery (file:///var/www/html/node_modules/@directus/api/dist/utils/get-ast-from-query.js:47:56)
              at ItemsService.readByQuery (file:///var/www/html/node_modules/@directus/api/dist/services/items.js:310:25)
              at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
              at async getExtensionsSettings (file:///var/www/html/node_modules/@directus/api/dist/extensions/lib/get-extensions-settings.js:16:22)
              at async ExtensionManager.load (file:///var/www/html/node_modules/@directus/api/dist/extensions/manager.js:138:39)
              at async ExtensionManager.initialize (file:///var/www/html/node_modules/@directus/api/dist/extensions/manager.js:122:13)
              at async createCli (file:///var/www/html/node_modules/@directus/api/dist/cli/index.js:23:9)
    }
`
Steps:
- npm install directus@10.7.0
- npx directus database migrate:latest`

**Comment by paescuj** [comment]
> I am also getting this issue. I delete the tables as stated above but then the same thing happens on the next time I migrate.
> 
> `err: { "type": "TypeError", "message": "Cannot read properties of undefined (reading 'primary')", "stack": TypeError: Cannot read properties of undefined (reading 'primary') at getASTFromQuery (file:///var/www/html/node_modules/@directus/api/dist/utils/get-ast-from-query.js:47:56) at ItemsService.readByQuery (file:///var/www/html/node_modules/@directus/api/dist/services/items.js:310:25) at process.processTicksAndRejections (node:internal/process/task_queues:95:5) at async getExtensionsSettings (file:///var/www/html/node_modules/@directus/api/dist/extensions/lib/get-extensions-settings.js:16:22) at async ExtensionManager.load (file:///var/www/html/node_modules/@directus/api/dist/extensions/manager.js:138:39) at async ExtensionManager.initialize (file:///var/www/html/node_modules/@directus/api/dist/extensions/manager.js:122:13) at async createCli (file:///var/www/html/node_modules/@directus/api/dist/cli/index.js:23:9) } ` Steps:
> 
>     * npm install directus@10.7.0
> 
>     * npx directus database migrate:latest`

This has been fixed in #20171 which has been released in v10.7.1.
